### PR TITLE
Kustomize API: implement a FieldSpec transformer using kyaml libraries

### PR DESCRIPTION
**Pull request by pwittrock**

### CODE REVIEW

**Comment by monopole** [comment]
Digging in now.  Please no more big changes for a bit :)
Hope to get this in today.

File: api/filters/fsslice/fieldspec.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+		// found the field -- set its value
+		return fltr.SetValue(obj)
+	}
+	switch obj.YNode().Kind {
+	case yaml.SequenceNode:
+		return fltr.seq(obj)
+	case yaml.MappingNode:
+		return fltr.field(obj)
+	}
+	// not found -- this might be an error since the type doesn't match
```
Comment (monopole) [inline api/filters/fsslice/fieldspec.go]:
```
return fmt.Err("unable to handle yaml node %v at path %v", obj.YNode, fltr.path)
```

or something like that

the code this replaces had very useful error messages for helping people find problems (in their cnfig and in our code)

File: api/filters/fsslice/fieldspec.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+		lookupField = yaml.LookupCreate(fltr.CreateKind, fieldName)
+	default:
+		// create the field if it is missing: must be a mapping node
+		lookupField = yaml.LookupCreate(yaml.MappingNode, fieldName)
+	}
+
+	// locate (or maybe create) the field
+	field, err := obj.Pipe(lookupField)
+	if err != nil || field == nil {
+		return err
```
Comment (monopole) [inline api/filters/fsslice/fieldspec.go]: due to recursion, please wrap err with more context.

return error.Wrap(err, "node=%v, lookupField=%s", ...)

File: api/filters/fsslice/fieldspec.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+// seq calls filter on all sequence elements
+func (fltr fieldSpecFilter) seq(obj *yaml.RNode) error {
+	if err := obj.VisitElements(func(node *yaml.RNode) error {
+		// recurse on each element -- re-allocating a fieldSpecFilter is
+		// not strictly required, but is more consistent with field
+		// and less likely to have side effects
+		// keep the entire path -- it does not contain parts for sequences
+		return fltr.filter(node)
+	}); err != nil {
+		return err
```
Comment (monopole) [inline api/filters/fsslice/fieldspec.go]:
```
return err.Wrap(err,'visit traversal in node %v', obj)
```
or something

File: plugin/builtin/namespacetransformer/NamespaceTransformer.go [hunk 4]
```diff
@@ -20,6 +22,10 @@ import (
 type plugin struct {
 	types.ObjectMeta `json:"metadata,omitempty" yaml:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
 	FieldSpecs       []types.FieldSpec `json:"fieldSpecs,omitempty" yaml:"fieldSpecs,omitempty"`
+
+	// LegacyMode can be set to true to use the kunstruct transformer version
+	// rather than the kyaml filter
+	LegacyMode bool
```
[the remark this answers is not in the document]
Comment (pwittrock) [reply plugin/builtin/namespacetransformer/NamespaceTransformer.go]: Would prefer legacy is the default behavior?

File: api/filters/filtersutil/filtersutil.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+import (
+	"encoding/json"
+	"strings"
+
+	"sigs.k8s.io/kustomize/api/resid"
+	"sigs.k8s.io/kustomize/kyaml/kio"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+// ApplyToJSON applies the filter to the json objects.
```
[the remark this answers is not in the document]
Comment (pwittrock) [reply api/filters/filtersutil/filtersutil.go]: done

File: api/filters/filtersutil/filtersutil.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+			return err
+		}
+	}
+
+	return nil
+}
+
+type marshalerUnmarshaler interface {
+	json.Unmarshaler
+	json.Marshaler
```
[the remark this answers is not in the document]
Comment (pwittrock) [reply api/filters/filtersutil/filtersutil.go]: :)

File: api/filters/fsslice/example_test.go [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+		CreateKind: yaml.ScalarNode,
+		SetValue:   fsslice.SetScalar("green"),
+		FsSlice: []types.FieldSpec{
+			{Path: "a/b", CreateIfNotPresent: true},
+		},
+	}
+
+	err := kio.Pipeline{
+		Inputs:  []kio.Reader{in},
+		Filters: []kio.Filter{kio.FilterAll(fltr)},
```
[the remark this answers is not in the document]
Comment (pwittrock) [reply api/filters/fsslice/example_test.go]: It is -- the fltr applies to a single node.  This is necessary to apply it to every node

File: api/filters/fsslice/doc.go [hunk 8]
```diff
@@ -0,0 +1,6 @@
+// Copyright 2020 The Kubernetes Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package fsslice contains a yaml.Filter to modify a resource using an
+// FsSlice to identify fields to be updated within the resource.
+package fsslice
```
[the remark this answers is not in the document]
Comment (pwittrock) [reply api/filters/fsslice/doc.go]: ack

File: api/filters/fsslice/fieldspec.go (hunk 2 above)
Comment (pwittrock) [reply api/filters/fsslice/fieldspec.go]: added the fieldName, the other parts will be added by at the entry point.

File: api/filters/filtersutil/filtersutil_test.go [hunk 9]
```diff
@@ -0,0 +1,30 @@
+package filtersutil_test
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"sigs.k8s.io/kustomize/api/filters/filtersutil"
+	"sigs.k8s.io/kustomize/api/resid"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
```
[the remark this answers is not in the document]
Comment (pwittrock) [reply api/filters/filtersutil/filtersutil_test.go]: done

File: api/filters/fsslice/fieldspec.go [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+import (
+	"strings"
+
+	"sigs.k8s.io/kustomize/api/types"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+// fieldSpecFilter applies a single fieldSpec to a single object
+// fieldSpecFilter stores internal state and should not be reused
+type fieldSpecFilter struct {
```
[the remark this answers is not in the document]
Comment (pwittrock) [reply api/filters/fsslice/fieldspec.go]: done

File: api/filters/fsslice/fieldspec.go (hunk 1 above)
Comment (pwittrock) [reply api/filters/fsslice/fieldspec.go]: done

File: api/filters/fsslice/fieldspec.go (hunk 3 above)
Comment (pwittrock) [reply api/filters/fsslice/fieldspec.go]: the node will be added by the entry point.  put the path in the message

**Comment by pwittrock** [comment]
PTAL

**Comment by mortent** [comment]
This looks good to me, but I'll leave it to @monopole to lgtm

File: api/filters/filtersutil/filtersutil.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (monopole) [reply api/filters/filtersutil/filtersutil.go]: I was not posing a question on how to maintain an API.

Do you think this should be maintained as part of the Kustomize API?

If not, please add a TODO to move it before the next kustomize API release.

File: plugin/builtin/namespacetransformer/NamespaceTransformer.go [hunk 11]
```diff
@@ -20,6 +22,10 @@ import (
 type plugin struct {
 	types.ObjectMeta `json:"metadata,omitempty" yaml:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
 	FieldSpecs       []types.FieldSpec `json:"fieldSpecs,omitempty" yaml:"fieldSpecs,omitempty"`
+
+	// YAMLSupport can be set to true to use the kyaml filter instead of the
+	// kunstruct transformer
+	YAMLSupport bool
```
Comment (monopole) [inline plugin/builtin/namespacetransformer/NamespaceTransformer.go]: add the json metadata please.

File: plugin/builtin/namespacetransformer/NamespaceTransformer.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (monopole) [reply plugin/builtin/namespacetransformer/NamespaceTransformer.go]: From the POV of the plugin, this is just another configuration value.

in https://github.com/kubernetes-sigs/kustomize/blob/master/plugin/builtin/namespacetransformer/NamespaceTransformer_test.go

the call 
```
	rm := th.LoadAndRunTransformer(...
```
takes two args: 1) the transform's config, 2) the thing to transform.

so the lowest effort way:

- copy the tests
- add the line `legacyMode :true` (or whatever you settle on for the field name and sense of the boolean) to the config section of half of the tests.

File: api/filters/filtersutil/filtersutil.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (pwittrock) [reply api/filters/filtersutil/filtersutil.go]: https://github.com/kubernetes-sigs/kustomize/pull/2297

**Comment by pwittrock** [comment]
Addressed remaining items in #173 2305

**Pull request merged**